### ISSUE TITLE
[FEA] Add feature hashing support

### ISSUE THREAD

**Issue by cmgreen210** [rapidsai/cudf#798]
**Is your feature request related to a problem? Please describe.**
I wish I could use cuDF to do [feature hashing](https://en.wikipedia.org/wiki/Feature_hashing) on the gpu similar to how scikit-learn's [FeatureHasher](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.FeatureHasher.html) handles string features: the feature name appended with the feature value are hashed to produce a column index of a sparse matrix. 

**Describe the solution you'd like**
Perhaps a method added to the `Series` class called `feature_hash`:

```python
def feature_hash(self, num_features, method='murmur'):
    """Perform feature hashing on values.

    Parameters
    ------------
    num_features: int
        Number of possible hash values, ``0, 1, ..., num_features-1``.
    method: str; optional
        Hashing method to use. Default 'murmur'. (See https://en.wikipedia.org/wiki/MurmurHash).

    Returns
    --------
    A new series with values in ``0, 1, ..., num_features - 1`` calculated from hashing series name
    with original value, ex: ``hash_func(self.name + str(val))``
    """
    ...
```

**Describe alternatives you've considered**
Feature hashing could be placed in another library such as cuML but I think this is such a common task it should be implemented here. 

One decision to be made is what to actually hash. You definitely **do not** want to hash just the value but some combination of the series name and the value: `"{name}_{value}"`.

**Additional context**
This functionality will be used extensively when using cuDF as an input to a deep learning frameworks.

**Comment by kkraus14** [comment]
@cmgreen210 How important is it for the hashing to be in a 0 --> n_features space? We currently having hashing that will map to (I believe) an int32 space that's used in groupbys and joins that we could expose as a function relatively easily, but unsure how difficult it would be to get it into a minimal space.

**Comment by kkraus14** [comment]
cc @jrhemstad @harrism

**Comment by nsakharnykh** [comment]
We have a cuDF function exposed that computes hashes for N columns:
https://github.com/rapidsai/cudf/blob/750118fb877fe245fa786826bfff54c97d675ce8/cpp/include/cudf/functions.h#L447

We can extend this function to take `num_features`, which then will build the hash table, and then extract the non-empty values, which will be mapped to 0 .. `num_features`. This is similar to building a hash table for joins + extracting the non-empty values like in the hash-based groupby, so I believe should be relatively easy to implement.

**Comment by cmgreen210** [comment]
> @cmgreen210 How important is it for the hashing to be in a 0 --> n_features space? We currently having hashing that will map to (I believe) an int32 space that's used in groupbys and joins that we could expose as a function relatively easily, but unsure how difficult it would be to get it into a minimal space.

@kkraus14 Very important since the hashed values will be used as lookup table indices. If you hash in an int32 space can you not just mod `num_features`?

**Comment by cmgreen210** [comment]
> We have a cuDF function exposed that computes hashes for N columns:
> [cudf/cpp/include/cudf/functions.h](https://github.com/rapidsai/cudf/blob/750118fb877fe245fa786826bfff54c97d675ce8/cpp/include/cudf/functions.h#L447)
> 
> Line 447 in [750118f](/rapidsai/cudf/commit/750118fb877fe245fa786826bfff54c97d675ce8)
> 
>  gdf_error gdf_hash(int num_cols, gdf_column **input, gdf_hash_func hash, gdf_column *output); 
> We can extend this function to take `num_features`, which then will build the hash table, and then extract the non-empty values, which will be mapped to 0 .. `num_features`. This is similar to building a hash table for joins + extracting the non-empty values like in the hash-based groupby, so I believe should be relatively easy to implement.

@nsakharnykh each element in the column, whether empty or not, should be hashed.

**Comment by cmgreen210** [comment]
@nsakharnykh For the purposes of feature hashing we do not care about collisions.

**Comment by muellren** [comment]
The existence of a good hash function h for uint32 does not imply that h(x) mod N is a good hash function for [0, N). When used in hash table one might just observe more collisions, which could be handled at a loss in performance. However, for feature hashing thing the quality of the resulting hash mapping has a direct impact on quality of the generated sketch. 

It gets easier if num_features is a power of two, then we can do XOR folding of the hash.

**Comment by kkraus14** [comment]
@cmgreen210 sure thing, this should likely eventually be moved down into the libcudf level, but implementing the expected functionality at the Python level in cuDF is a great start.

**Comment by cmgreen210** [comment]
> @cmgreen210 sure thing, this should likely eventually be moved down into the libcudf level, but implementing the expected functionality at the Python level in cuDF is a great start.

I will work on the `libcudf` component as well.

**Comment by cmgreen210** [comment]
@nsakharnykh After looking over the code I think adding a new function to `cudf/cpp/include/cudf/functions.h` rather than extending `gdf_hash` is the way to go for a few reasons. First, this transformation will only be done on a single column at a time so instead of `int num_cols, gdf_column **input` all we need is a `gdf_column *input`. Second, we need to hash the name of the column with the column values. We can accomplish this by accessing the column name at `input->col_name`, hashing this value, and then for each element in the column combine its value with the hashed column name using [hash_combine](https://github.com/rapidsai/cudf/blob/b7356d9204acc5922a54fd7814c105390aaaafa5/cpp/src/hash/hash_functions.cuh#L73). 

So, we could add a function with signature:
```python
gdf_error gdf_feature_hash(gdf_column *input, gdf_hash_func hash, gdf_size_type num_features, gdf_column *output);
```

What do you think? cc @kkraus14

**Comment by muellren** [comment]
@cmgreen210 As suggestion make the hashing of the column name and optional in libcuml that could be turned off (e.g., add a boolean). 

> I wish I could use cuDF to do feature hashing on the gpu much like scikit-learn's FeatureHasher.

Just for clarity, the intended feature_hash function is not meant to perform the same operation as scikits [FeatureHasher](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.FeatureHasher.html#sklearn.feature_extraction.FeatureHasher.transform), correct? The FeatureHasher returns a (M x num_features) CSR matrix. It takes an iterable value x, computes the murmur3 hash h(x), determines the column index as (h(x) % num_features),  multiplies the value (if x is dict) or 1 with (-1)*MSB of h(x) and adds it to the column.

**Comment by cmgreen210** [comment]
> @cmgreen210 As suggestion make the hashing of the column name and optional in libcuml that could be turned off (e.g., add a boolean).
> 
> > I wish I could use cuDF to do feature hashing on the gpu much like scikit-learn's FeatureHasher.
> 
> Just for clarity, the intended feature_hash function is not meant to perform the same operation as scikits [FeatureHasher](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.FeatureHasher.html#sklearn.feature_extraction.FeatureHasher.transform), correct? The FeatureHasher returns a (M x num_features) CSR matrix. It takes an iterable value x, computes the murmur3 hash h(x), determines the column index as (h(x) % num_features), multiplies the value (if x is dict) or 1 with (-1)*MSB of h(x) and adds it to the column.

@muellren Right it's not meant to do the same as scikit. It's intended to essentially perform something similar to how scikit handles string features.

Making the column name hashing optional is ok with me.
